File: scrape_axiom.py
import sys
import time
import re
from playwright.sync_api import sync_playwright

def scrape_axiom():
    print("=" * 60)
    print("🚀 INICIANDO SCRAPER DE AXIOM.TRADE (SOLANA PULSE) - PYTHON...")
    print("=" * 60)
    print("💡 Abrindo o navegador (modo visível para contornar Cloudflare)...")
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=False)
        context = browser.new_context(
            viewport={"width": 1280, "height": 800},
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        )
        
        page = context.new_page()
        
        print("🔗 Acessando: https://axiom.trade/pulse?chain=sol")
        try:
            page.goto("https://axiom.trade/pulse?chain=sol", wait_until="commit", timeout=60000)
        except Exception as e:
            print(f"❌ Erro ao abrir a página: {e}")
            browser.close()
            return

        print("\n⏳ Aguardando a página carregar e o Cloudflare ser resolvido...")
        print("👉 Se aparecer um desafio do Cloudflare (Turnstile/Captcha), clique nele para passar!")
        
        # Aguarda 15 segundos para dar tempo do Cloudflare validar
        for i in range(15, 0, -1):
            sys.stdout.write(f"\rAguardando carregamento... {i}s restantes ")
            sys.stdout.flush()
            time.sleep(1)
        print("\n\n🔍 Iniciando verificação de login...")

        # --- FLUXO DE LOGIN AUTOMÁTICO ---
        try:
            login_link = page.locator('text="Login"').first
            if login_link.is_visible():
                print("🔑 Botão de Login detectado! Realizando autenticação automática...")
                login_link.click()
                
                # Aguarda o modal de login terminar de renderizar e o input ficar visível
                print("⏳ Aguardando campos de formulário ficarem visíveis...")
                email_input = page.locator('input[type="email"], input[placeholder*="email"], input[placeholder*="Email"], input[placeholder*="Enter email"]').first
                pass_input = page.locator('input[type="password"], input[placeholder*="password"], input[placeholder*="Password"], input[placeholder*="Enter password"]').first
                
                email_input.wait_for(state="visible", timeout=15000)

                print("📧 Inserindo credenciais de login...")
                email_input.fill('tobizinho12345678@gmail.com')
                pass_input.fill('laisa1.9')
                
                print("🔘 Clicando no botão de confirmação de Login...")
                submit_btn = page.locator('button:has-text("Login")').last
                submit_btn.click()
                
                print("⏳ Aguardando redirecionamento após login (10s)...")
                page.wait_for_timeout(10000)
            else:
                print("ℹ️ Botão de Login inicial não encontrado. Você já pode estar logado ou a página foi direto para o painel.")
        except Exception as e:
            print(f"⚠️ Não foi possível automatizar o login ou já estava logado: {e}")

        # --- SCREENSHOT DE SEGURANÇA ---
        try:
            page.screenshot(path="axiom_pulse_screenshot_py.png")
            print("📸 Screenshot da página salvo como 'axiom_pulse_screenshot_py.png'")
        except Exception as e:
            print(f"⚠️ Não foi possível tirar screenshot: {e}")

        # --- EXTRAÇÃO E PROCESSAMENTO DOS DADOS ---
        print("\n🔍 Capturando dados da tabela Pulse...")
        try:
            body_text = page.locator("body").inner_text()
            lines = [line.strip() for line in body_text.split("\n") if line.strip()]
            
            print("\n" + "=" * 50)
            print("🎯 PROCESSANDO TOKENS DETECTADOS (PULSE SOLANA):")
            print("=" * 50)
            
            detected_tokens = []
            current_token = None
            
            i = 0
            while i < len(lines):
                line = lines[i]
                
                if re.match(r'^\d+[smhd]\s+ago$', line):
                    if current_token and current_token.get("ticker"):
                        detected_tokens.append(current_token)
                        
                    current_token = {
                        "age": line,
                        "ticker": "N/A",
                        "name": "N/A",
                        "market_cap": "N/A",
                        "dev_hold": "N/A",
                        "progress": "N/A"
                    }
                    
                    if i + 1 < len(lines): current_token["ticker"] = lines[i + 1]
                    if i + 2 < len(lines): current_token["name"] = lines[i + 2]
                    if i + 3 < len(lines): current_token["market_cap"] = lines[i + 3]
                    if i + 4 < len(lines): current_token["dev_hold"] = lines[i + 4]
                    if i + 5 < len(lines): current_token["progress"] = lines[i + 5]
                    
                    i += 5
                i += 1
                
            if current_token and current_token.get("ticker") and current_token["ticker"] != "N/A":
                detected_tokens.append(current_token)

            if detected_tokens:
                print(f"✅ Sucesso! Detectamos {len(detected_tokens)} moedas ativas na Pulse:")
                print("-" * 80)
                for idx, t in enumerate(detected_tokens):
                    ticker = t["ticker"].ljust(8)
                    mcap = t["market_cap"].ljust(8)
                    prog = t["progress"].ljust(6)
                    dev = t["dev_hold"].ljust(6)
                    age = t["age"]
                    print(f"✨ Coin #{idx+1:02d} | Ticker: {ticker} | MCap: {mcap} | Curva: {prog} | Dev Hold: {dev} | Idade: {age}")
                print("-" * 80)
            else:
                print("⚠️ Nenhum padrão de moeda foi extraído com sucesso.")
                print("Dica: Verifique se a página carregou completamente ou se a conta do usuário precisa resolver algum desafio manual.")
                
        except Exception as e:
            print(f"❌ Erro ao extrair textos: {e}")

        print("\n" + "=" * 60)
        print("🏁 Finalizando o scraper. Fechando navegador...")
        print("=" * 60)
        browser.close()

if __name__ == "__main__":
    scrape_axiom()

File: scrape_axiom.js
const { chromium } = require('playwright');
const fs = require('fs');

async function scrapeAxiom() {
  console.log("=".repeat(60));
  console.log("🚀 INICIANDO SCRAPER DE AXIOM.TRADE (SOLANA PULSE) - NODE.JS...");
  console.log("=".repeat(60));
  console.log("💡 Abrindo o navegador (modo visível para contornar Cloudflare)...");

  // Inicia o navegador Chromium visível
  const browser = await chromium.launch({ headless: false });
  const context = await browser.newContext({
    viewport: { width: 1280, height: 800 },
    userAgent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
  });

  const page = await context.newPage();

  console.log("🔗 Acessando: https://axiom.trade/pulse?chain=sol");
  try {
    // Usamos waitUntil: "commit" para não travar indefinidamente em websockets/analytics
    await page.goto("https://axiom.trade/pulse?chain=sol", { waitUntil: "commit", timeout: 60000 });
  } catch (err) {
    console.error(`❌ Erro ao abrir a página: ${err.message}`);
    await browser.close();
    return;
  }

  console.log("\n⏳ Aguardando a página carregar e o Cloudflare ser resolvido...");
  console.log("👉 Se aparecer um desafio do Cloudflare (Turnstile/Captcha), clique nele para passar!");
  
  // Aguarda 15 segundos para dar tempo de resolver o Cloudflare e carregar a página
  for (let i = 15; i > 0; i--) {
    process.stdout.write(`\rAguardando carregamento... ${i}s restantes `);
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  console.log("\n\n🔍 Iniciando verificação de login...");

  // --- FLUXO DE LOGIN AUTOMÁTICO ---
  try {
    const loginLink = page.locator('text="Login"').first();
    if (await loginLink.isVisible()) {
      console.log("🔑 Botão de Login detectado! Realizando autenticação automática...");
      await loginLink.click();
      
      // Aguarda o modal de login terminar de renderizar e o input ficar visível
      console.log("⏳ Aguardando campos de formulário ficarem visíveis...");
      const emailInput = page.locator('input[type="email"], input[placeholder*="email"], input[placeholder*="Email"], input[placeholder*="Enter email"]').first();
      const passInput = page.locator('input[type="password"], input[placeholder*="password"], input[placeholder*="Password"], input[placeholder*="Enter password"]').first();
      
      await emailInput.waitFor({ state: 'visible', timeout: 15000 });

      console.log("📧 Inserindo credenciais de login...");
      await emailInput.fill('tobizinho12345678@gmail.com');
      await passInput.fill('laisa1.9');
      
      console.log("🔘 Clicando no botão de confirmação de Login...");
      const submitBtn = page.locator('button:has-text("Login")').last();
      await submitBtn.click();
      
      console.log("⏳ Aguardando redirecionamento após login (10s)...");
      await page.waitForTimeout(10000);
    } else {
      console.log("ℹ️ Botão de Login inicial não encontrado. Você já pode estar logado ou a página foi direto para o painel.");
    }
  } catch (err) {
    console.warn(`⚠️ Não foi possível automatizar o login ou já estava logado: ${err.message}`);
  }

  // --- SCREENSHOT DE SEGURANÇA ---
  try {
    await page.screenshot({ path: "axiom_pulse_screenshot_node.png" });
    console.log("📸 Screenshot salvo como 'axiom_pulse_screenshot_node.png'");
  } catch (err) {
    console.warn(`⚠️ Não foi possível tirar screenshot: ${err.message}`);
  }

  // --- EXTRAÇÃO E PROCESSAMENTO DOS DADOS ---
  console.log("\n🔍 Capturando dados da tabela Pulse...");
  try {
    const bodyText = await page.locator("body").innerText();
    const lines = bodyText.split("\n").map(l => l.trim()).filter(l => l.length > 0);
    
    console.log("\n" + "=".repeat(50));
    console.log("🎯 PROCESSANDO TOKENS DETECTADOS (PULSE SOLANA):");
    console.log("=".repeat(50));

    const detectedTokens = [];
    let currentToken = null;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      
      // Encontra a linha de idade como âncora (ex: "15s ago", "2m ago", "1h ago")
      if (/^\d+[smhd]\s+ago$/.test(line)) {
        if (currentToken && currentToken.ticker) {
          detectedTokens.push(currentToken);
        }
        
        currentToken = {
          age: line,
          ticker: "N/A",
          name: "N/A",
          marketCap: "N/A",
          devHold: "N/A",
          progress: "N/A"
        };
        
        if (i + 1 < lines.length) currentToken.ticker = lines[i + 1];
        if (i + 2 < lines.length) currentToken.name = lines[i + 2];
        if (i + 3 < lines.length) currentToken.marketCap = lines[i + 3];
        if (i + 4 < lines.length) currentToken.devHold = lines[i + 4];
        if (i + 5 < lines.length) currentToken.progress = lines[i + 5];
        
        i += 5;
      }
    }
    
    if (currentToken && currentToken.ticker && currentToken.ticker !== "N/A") {
      detectedTokens.push(currentToken);
    }

    if (detectedTokens.length > 0) {
      console.log(`✅ Sucesso! Detectamos ${detectedTokens.length} moedas ativas na Pulse:`);
      console.log("-".repeat(80));
      detectedTokens.forEach((t, idx) => {
        console.log(`✨ Coin #${(idx+1).toString().padStart(2, '0')} | Ticker: ${t.ticker.padEnd(8)} | MCap: ${t.marketCap.padEnd(8)} | Curva: ${t.progress.padEnd(6)} | Dev Hold: ${t.devHold.padEnd(6)} | Idade: ${t.age}`);
      });
      console.log("-".repeat(80));
    } else {
      console.log("⚠️ Nenhum padrão de moeda foi extraído com sucesso.");
      console.log("Dica: Verifique se a página carregou completamente ou se a conta do usuário precisa resolver algum desafio manual.");
    }

  } catch (err) {
    console.error(`❌ Erro ao extrair textos da página: ${err.message}`);
  }

  console.log("\n" + "=".repeat(60));
  console.log("🏁 Finalizando o scraper. Fechando navegador...");
  console.log("=".repeat(60));
  await browser.close();
}

scrapeAxiom();
